lcls_tools/superconducting/sc_cavity.py

```diff
@@ -0,0 +1,898 @@
+        return self._calc_probe_q_pv_obj
+
+    def calculate_probe_q(self):
+        self.calc_probe_q_pv_obj.put(1)
+
+    def set_chirp_range(self, offset: int):
+        offset = abs(offset)
+        print(f"Setting chirp range for {self} to +/- {offset} Hz")
+        self.chirp_freq_start = -offset
+        self.chirp_freq_stop = offset
+        print(f"Chirp range set for {self}")
+
+    @property
+    def rf_state_pv_obj(self) -> PV:
+        if not self._rf_state_pv_obj:
+            self._rf_state_pv_obj = PV(self.rf_state_pv)
+        return self._rf_state_pv_obj
+
+    @property
+    def rf_state(self):
+        """This property is read only"""
+        return self.rf_state_pv_obj.get()
+
+    @property
+    def is_on(self):
+        return self.rf_state == 1
+
+    def delta_piezo(self):
+        delta_volts = self.piezo.voltage - utils.PIEZO_CENTER_VOLTAGE
+        delta_hz = delta_volts * utils.PIEZO_HZ_PER_VOLT
+        print(f"{self} piezo detune: {delta_hz}")
+        return delta_hz if not self.cryomodule.is_harmonic_linearizer else -delta_hz
+
+    def move_to_resonance(self, reset_signed_steps=False, use_sela=False):
+        def delta_detune():
+            return self.detune
+
+        self.setup_tuning(use_sela=use_sela)
+        print(f"Tuning {self} to resonance in " + ("SELA" if use_sela else "chirp"))
+        self._auto_tune(
+            delta_hz_func=delta_detune,
+            tolerance=(500 if self.cryomodule.is_harmonic_linearizer else 50),
+            reset_signed_steps=reset_signed_steps,
+        )
+
+        if use_sela:
+            print(f"Centering {self} piezo")
+            self._auto_tune(
+                delta_hz_func=self.delta_piezo,
+                tolerance=100,
+                reset_signed_steps=False,
+            )
+
+        self.tune_config_pv_obj.put(utils.TUNE_CONFIG_RESONANCE_VALUE)
+
+    @property
+    def detune_best_pv_obj(self) -> PV:
+        if not self._detune_best_pv_obj:
+            self._detune_best_pv_obj = PV(self.detune_best_pv)
+        return self._detune_best_pv_obj
+
+    @property
+    def detune_chirp_pv_obj(self) -> PV:
+        if not self._detune_chirp_pv_obj:
+            self._detune_chirp_pv_obj = PV(self.detune_chirp_pv)
+        return self._detune_chirp_pv_obj
+
+    @property
+    def detune_best(self):
+        return self.detune_best_pv_obj.get()
+
+    @property
+    def detune_chirp(self):
+        return self.detune_chirp_pv_obj.get()
+
+    @property
+    def detune(self):
+        if self.rf_mode == utils.RF_MODE_CHIRP:
+            return self.detune_chirp
+        else:
+            return self.detune_best
+
+    @property
+    def detune_invalid(self) -> bool:
+        if self.rf_mode == utils.RF_MODE_CHIRP:
+            return self.detune_chirp_pv_obj.severity == EPICS_INVALID_VAL
+        else:
+            return self.detune_best_pv_obj.severity == EPICS_INVALID_VAL
+
+    def _auto_tune(
+        self,
```

Comment:
It's called in `move_to_resonance` which I think is expected to be used every time you want to run RF in a cavity because I've had situations where a cavity is misbehaving because the detune ran away without me noticing